extensions/oidc-devservices-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesProcessor.java

```diff
@@ -0,0 +1,395 @@
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.DevServicesNativeConfigResultBuildItem;
+import io.quarkus.deployment.builditem.LaunchModeBuildItem;
+import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
+import io.quarkus.deployment.builditem.ServiceStartBuildItem;
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.configuration.ConfigUtils;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.mutiny.core.buffer.Buffer;
+import io.vertx.mutiny.ext.web.client.WebClient;
+
+public class KeycloakDevServicesProcessor {
+    private static final Logger LOG = Logger.getLogger(KeycloakDevServicesProcessor.class);
+
+    private static final String CONFIG_PREFIX = "quarkus.oidc.";
+    private static final String AUTH_SERVER_URL_CONFIG_KEY = CONFIG_PREFIX + "auth-server-url";
+    private static final String APPLICATION_TYPE_CONFIG_KEY = CONFIG_PREFIX + "application-type";
+    private static final String CLIENT_ID_CONFIG_KEY = CONFIG_PREFIX + "client-id";
+    private static final String CLIENT_SECRET_CONFIG_KEY = CONFIG_PREFIX + "credentials.secret";
+    private static final String KEYCLOAK_URL_KEY = "keycloak.url";
+
+    private static final int KEYCLOAK_EXPOSED_PORT = 8080;
+    private static final String KEYCLOAK_DOCKER_REALM_PATH = "/tmp/realm.json";
+    private static final String KEYCLOAK_USER_PROP = "KEYCLOAK_USER";
+    private static final String KEYCLOAK_PASSWORD_PROP = "KEYCLOAK_PASSWORD";
+    private static final String KEYCLOAK_VENDOR_PROP = "DB_VENDOR";
+    private static final String KEYCLOAK_IMPORT_PROP = "KEYCLOAK_IMPORT";
+    private static final String KEYCLOAK_ADMIN_USER = "admin";
+    private static final String KEYCLOAK_ADMIN_PASSWORD = "admin";
+    private static final String KEYCLOAK_DB_VENDOR = "H2";
+    private static volatile List<Closeable> closeables;
+    private static volatile DevServicesConfig capturedDevServicesConfiguration;
+    private static volatile boolean first = true;
+    private final IsDockerWorking isDockerWorking = new IsDockerWorking(true);
+
+    KeycloakBuildTimeConfig config;
+
+    @BuildStep(onlyIfNot = IsNormal.class)
+    public DevServicesConfigBuildItem startOIDCContainer(LaunchModeBuildItem launchMode,
+            BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfiguration,
+            BuildProducer<ServiceStartBuildItem> serviceStartBuildItemBuildProducer,
+            BuildProducer<DevServicesNativeConfigResultBuildItem> devServices) {
+
+        DevServicesConfig currentDevServicesConfiguration = config.devservices;
+        // figure out if we need to shut down and restart any existing Keycloak container
+        // if not and the Keycloak container have already started we just return
+        if (closeables != null) {
+            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.TEST;
```

Comment:
We should not be restarting for test mode, unless there has been a config change. This makes continuous testing super slow if you need to restart the container every time.

Reply:
Hi Stuart, @stuartwdouglas, this is how it is done now:
```
// This step runs for both Test and Developement modes
@BuildStep(onlyIfNot = IsNormal.class)
public DevServicesConfigBuildItem startKeycloakContainer(...) {

        DevServicesConfig currentDevServicesConfiguration = config.devservices;
        // figure out if we need to shut down and restart any existing Keycloak container
        // if not and the Keycloak container have already started we just return
        if (closeables != null) {
            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.DEVELOPMENT
                    && !currentDevServicesConfiguration.equals(capturedDevServicesConfiguration);

            if (!restartRequired) {
                return null;
            }
            // close all the closeables
            // restart
            StartResult startResult = startContainer(currentDevServicesConfiguration);
           ...
}
```

I've copied `@BuildStep(onlyIfNot = IsNormal.class)` from one of the other dev services processors.

Should it be `@BuildStep(onlyIf = IsDevelopment.class)`, i.e, when running the tests we should expect a test specific container starting with the test resource manager ?

Reply:
I think I get it - so the container won't start if `quarkus.oidc.auth-server-url' is set - which is a prerequisite for running the tests so the above should be fine (I hope :-) )